File: lang/go/src/fun/vac/data_structures/LinkedQueue/LinkedQueue.go
package LinkedQueue

import "errors"

type Node struct {
    data int
    next *Node
}

type LinkedQueue struct {
    head *Node
    tail *Node
    length int
}

func New() *LinkedQueue {
    return &LinkedQueue{head: nil, tail: nil, length: 0}
}

func (queue *LinkedQueue) Size() int {
    return queue.length
}

func (queue *LinkedQueue) IsEmpty() bool {
    return queue.length == 0
}

func (queue *LinkedQueue) Peek() int {
    if queue.length == 0 {
        panic(errors.New("[PANIC - NoSuchElement]"))
    }

    return queue.head.data
}

func (queue *LinkedQueue) Enqueue(element int) {
    node := &Node{data: element, next: nil}

    if queue.length == 0 {
        queue.head = node
    } else {
        queue.tail.next = node
    }

    queue.tail = node

    queue.length += 1
}

func (queue *LinkedQueue) Dequeue() {
    if queue.length == 0 {
        panic(errors.New("[PANIC - NoSuchElement]"))
    }

    target := queue.head

    if queue.length == 1 {
        queue.head = nil
        queue.tail = nil
    } else {
        queue.head = target.next
    }

    target.data = int(0)

    queue.length -= 1
}

File: archive/nodejs/muse/data_structures/LinkedQueue.js
class Node {
  constructor(element) {
    this.data = element;
    this.next = null;
  }
}

export class LinkedQueue {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  size() {
    return this.length;
  }

  isEmpty() {
    return this.length === 0;
  }

  peek() {
    if (this.length === 0) {
      throw new Error('[PANIC - NoSuchElement]');
    }

    return this.head.data;
  }

  enqueue(element) {
    const node = new Node(element);

    if (this.length === 0) {
      this.head = node;
    } else {
      this.tail.next = node;
    }

    this.tail = node;

    this.length++;
  }

  dequeue() {
    if (this.length === 0) {
      throw new Error('[PANIC - NoSuchElement]');
    }

    const target = this.head;

    if (this.length === 1) {
      this.head = null;
      this.tail = null;
    } else {
      this.head = target.next;
    }

    target.data = null;

    this.length--;
  }
}
